File: modules/matterhorn-common/src/main/java/org/opencastproject/util/HttpUtil.java
/**
 *  Copyright 2009, 2010 The Regents of the University of California
 *  Licensed under the Educational Community License, Version 2.0
 *  (the "License"); you may not use this file except in compliance
 *  with the License. You may obtain a copy of the License at
 *
 *  http://www.osedu.org/licenses/ECL-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an "AS IS"
 *  BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 *  or implied. See the License for the specific language governing
 *  permissions and limitations under the License.
 *
 */
package org.opencastproject.util;

import static org.opencastproject.util.EqualsUtil.eq;
import static org.opencastproject.util.data.Collections.list;
import static org.opencastproject.util.data.Either.left;
import static org.opencastproject.util.data.Either.right;
import static org.opencastproject.util.data.Monadics.mlist;
import static org.opencastproject.util.data.Prelude.sleep;
import static org.opencastproject.util.data.functions.Misc.chuck;

import org.opencastproject.security.api.TrustedHttpClient;
import org.opencastproject.security.api.TrustedHttpClient.RequestRunner;
import org.opencastproject.util.data.Collections;
import org.opencastproject.util.data.Either;
import org.opencastproject.util.data.Function;
import org.opencastproject.util.data.Tuple;

import org.apache.commons.io.IOUtils;
import org.apache.http.Header;
import org.apache.http.HttpEntityEnclosingRequest;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.NameValuePair;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpHead;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.message.BasicNameValuePair;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URLEncoder;
import java.util.List;

/** Functions to support Apache httpcomponents and HTTP related operations in general. */

/** Functions to support Apache httpcomponents. */
public final class HttpUtil {
  private HttpUtil() {
  }

  public static HttpPost post(NameValuePair... formParams) {
    final HttpPost post = new HttpPost();
    setFormParams(post, formParams);
    return post;
  }

  public static HttpPost post(String uri, NameValuePair... formParams) {
    final HttpPost post = new HttpPost(uri);
    setFormParams(post, formParams);
    return post;
  }

  public static HttpPost post(String uri, List<NameValuePair> formParams) {
    final HttpPost post = new HttpPost(uri);
    setFormParams(post, Collections.toArray(NameValuePair.class, formParams));
    return post;
  }

  public static HttpGet get(String path, Tuple<String, String>... queryParams) {
    final String url = mlist(path, mlist(queryParams).map(new Function<Tuple<String, String>, String>() {
      @Override
      public String apply(Tuple<String, String> a) {
        try {
          return a.getA() + "=" + URLEncoder.encode(a.getB(), "UTF-8");
        } catch (UnsupportedEncodingException e) {
          return chuck(e);
        }
      }
    }).mkString("&")).mkString("?");
    return new HttpGet(url);
  }

  public static String path(String... path) {
    return UrlSupport.concat(path);
  }

  private static void setFormParams(HttpEntityEnclosingRequest r, NameValuePair[] formParams) {
    final List<NameValuePair> params = list(formParams);
    try {
      r.setEntity(new UrlEncodedFormEntity(params, "UTF-8"));
    } catch (UnsupportedEncodingException e) {
      chuck(e);
    }
  }

  public static NameValuePair param(String name, String value) {
    return new BasicNameValuePair(name, value);
  }

  public static NameValuePair param(Tuple<String, String> p) {
    return new BasicNameValuePair(p.getA(), p.getB());
  }

  public static final Function<Tuple<String, String>, NameValuePair> param_ = new Function<Tuple<String, String>, NameValuePair>() {
    @Override
    public NameValuePair apply(Tuple<String, String> p) {
      return param(p);
    }
  };

  public static final Function<HttpResponse, Integer> getStatusCode = new Function<HttpResponse, Integer>() {
    @Override
    public Integer apply(HttpResponse response) {
      return response.getStatusLine().getStatusCode();
    }
  };

  /**
   * Return the content of the response as a string if its status code equals one of the given statuses. Throw an
   * exception on an unexpected status.
   * <p/>
   * Function composition of {@link #getContentFn} and {@link #expect(int...)}.
   */
  public static Function<HttpResponse, String> getContentOn(final int... status) {
    return getContentFn.o(expect(status));
  }

  public static String getContentOn(final RequestRunner<String> runner, final int... status) {
    final Either<Exception, String> res = runner.run(getContentOn(status));
    if (res.isRight()) {
      return res.right().value();
    } else {
      return chuck(res.left().value());
    }
  }

  /** Return the content of the response as a string. */
  public static final Function<HttpResponse, String> getContentFn = new Function.X<HttpResponse, String>() {
    @Override
    public String xapply(HttpResponse httpResponse) throws Exception {
      final Header h = httpResponse.getEntity().getContentEncoding();
      if (h != null) {
        return IOUtils.toString(httpResponse.getEntity().getContent(), h.getValue());
      } else {
        return IOUtils.toString(httpResponse.getEntity().getContent());
      }
    }
  };

  /** Return the response if its status code equals one of the given statuses or throw an exception. */
  public static Function<HttpResponse, HttpResponse> expect(final int... status) {
    return new Function.X<HttpResponse, HttpResponse>() {
      @Override
      public HttpResponse xapply(HttpResponse response) {
        final int sc = response.getStatusLine().getStatusCode();
        for (int s : status) {
          if (sc == s)
            return response;
        }
        String responseBody;
        try {
          responseBody = IOUtils.toString(response.getEntity().getContent());
        } catch (IOException e) {
          responseBody = "";
        }
        throw new RuntimeException("Returned status " + sc + " does not match any of the expected codes. "
                + responseBody);
      }
    };
  }

  /** Get the value or throw the exception. */
  public static <A> A getOrError(Either<Exception, A> response) {
    if (response.isRight()) {
      return response.right().value();
    } else {
      return chuck(response.left().value());
    }
  }

  public static boolean isOk(HttpResponse res) {
    return res.getStatusLine().getStatusCode() == HttpStatus.SC_OK;
  }

  /**
   * Wait for a certain status of a resource.
   *
   * @return either an exception or the status code of the last http response
   */
  public static Either<Exception, Integer> waitForResource(final TrustedHttpClient http, final URI resourceUri,
          final int expectedStatus, final long timeout, final long pollingInterval) {
    long now = 0L;
    while (true) {
      final HttpHead head = new HttpHead(resourceUri);
      final Either<Exception, Integer> result = http.<Integer> run(head).apply(getStatusCode);
      for (final Integer status : result.right()) {
        if (eq(status, expectedStatus) || now >= timeout) {
          return right(status);
        } else if (now < timeout) {
          if (!sleep(pollingInterval)) {
            return left(new Exception("Interrupted"));
          } else {
            now = now + pollingInterval;
          }
        }
      }
      for (Exception e : result.left()) {
        return left(e);
      }
    }
  }
}

File: modules/common/src/main/java/org/opencastproject/util/HttpUtil.kt
/**
 * Licensed to The Apereo Foundation under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 *
 * The Apereo Foundation licenses this file to you under the Educational
 * Community License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License
 * at:
 *
 * http://opensource.org/licenses/ecl2.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package org.opencastproject.util

import org.opencastproject.util.EqualsUtil.eq
import org.opencastproject.util.data.Collections.list
import org.opencastproject.util.data.Either.left
import org.opencastproject.util.data.Either.right
import org.opencastproject.util.data.Monadics.mlist
import org.opencastproject.util.data.Prelude.sleep
import org.opencastproject.util.data.functions.Misc.chuck

import org.opencastproject.security.api.TrustedHttpClient
import org.opencastproject.security.api.TrustedHttpClient.RequestRunner
import org.opencastproject.util.data.Collections
import org.opencastproject.util.data.Either
import org.opencastproject.util.data.Function
import org.opencastproject.util.data.Tuple

import org.apache.commons.io.IOUtils
import org.apache.http.Header
import org.apache.http.HttpEntityEnclosingRequest
import org.apache.http.HttpResponse
import org.apache.http.HttpStatus
import org.apache.http.NameValuePair
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.HttpGet
import org.apache.http.client.methods.HttpHead
import org.apache.http.client.methods.HttpPost
import org.apache.http.message.BasicNameValuePair

import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.URI
import java.net.URLEncoder

/** Functions to support Apache httpcomponents and HTTP related operations in general.  */

/** Functions to support Apache httpcomponents.  */
object HttpUtil {

    val param_: Function<Tuple<String, String>, NameValuePair> = object : Function<Tuple<String, String>, NameValuePair>() {
        override fun apply(p: Tuple<String, String>): NameValuePair {
            return param(p)
        }
    }

    val getStatusCode: Function<HttpResponse, Int> = object : Function<HttpResponse, Int>() {
        override fun apply(response: HttpResponse): Int {
            return response.statusLine.statusCode
        }
    }

    /** Return the content of the response as a string.  */
    val getContentFn: Function<HttpResponse, String> = object : Function.X<HttpResponse, String>() {
        @Throws(Exception::class)
        public override fun xapply(httpResponse: HttpResponse): String {
            val h = httpResponse.entity.contentEncoding
            return if (h != null) {
                IOUtils.toString(httpResponse.entity.content, h.value)
            } else {
                IOUtils.toString(httpResponse.entity.content)
            }
        }
    }

    fun post(vararg formParams: NameValuePair): HttpPost {
        val post = HttpPost()
        setFormParams(post, formParams)
        return post
    }

    fun post(uri: String, vararg formParams: NameValuePair): HttpPost {
        val post = HttpPost(uri)
        setFormParams(post, formParams)
        return post
    }

    fun post(uri: String, formParams: List<NameValuePair>): HttpPost {
        val post = HttpPost(uri)
        setFormParams(post, Collections.toArray(NameValuePair::class.java, formParams))
        return post
    }

    operator fun get(path: String, vararg queryParams: Tuple<String, String>): HttpGet {
        val url = mlist(path, mlist(*queryParams).map(object : Function<Tuple<String, String>, String>() {
            override fun apply(a: Tuple<String, String>): String {
                try {
                    return a.a + "=" + URLEncoder.encode(a.b, "UTF-8")
                } catch (e: UnsupportedEncodingException) {
                    return chuck(e)
                }

            }
        }).mkString("&")).mkString("?")
        return HttpGet(url)
    }

    fun path(vararg path: String): String {
        return UrlSupport.concat(*path)
    }

    private fun setFormParams(r: HttpEntityEnclosingRequest, formParams: Array<NameValuePair>) {
        val params = list(*formParams)
        try {
            r.entity = UrlEncodedFormEntity(params, "UTF-8")
        } catch (e: UnsupportedEncodingException) {
            chuck<Any>(e)
        }

    }

    fun param(name: String, value: String): NameValuePair {
        return BasicNameValuePair(name, value)
    }

    fun param(p: Tuple<String, String>): NameValuePair {
        return BasicNameValuePair(p.a, p.b)
    }

    /**
     * Return the content of the response as a string if its status code equals one of the given statuses. Throw an
     * exception on an unexpected status.
     *
     *
     * Function composition of [.getContentFn] and [.expect].
     */
    fun getContentOn(vararg status: Int): Function<HttpResponse, String> {
        return getContentFn.o(expect(*status))
    }

    fun getContentOn(runner: RequestRunner<String>, vararg status: Int): String {
        val res = runner.run(getContentOn(*status))
        return if (res.isRight) {
            res.right().value()
        } else {
            chuck(res.left().value())
        }
    }

    /** Return the response if its status code equals one of the given statuses or throw an exception.  */
    fun expect(vararg status: Int): Function<HttpResponse, HttpResponse> {
        return object : Function.X<HttpResponse, HttpResponse>() {
            public override fun xapply(response: HttpResponse): HttpResponse {
                val sc = response.statusLine.statusCode
                for (s in status) {
                    if (sc == s)
                        return response
                }
                var responseBody: String
                try {
                    responseBody = IOUtils.toString(response.entity.content)
                } catch (e: IOException) {
                    responseBody = ""
                }

                throw RuntimeException("Returned status " + sc + " does not match any of the expected codes. "
                        + responseBody)
            }
        }
    }

    /** Get the value or throw the exception.  */
    fun <A> getOrError(response: Either<Exception, A>): A {
        return if (response.isRight) {
            response.right().value()
        } else {
            chuck(response.left().value())
        }
    }

    fun isOk(res: HttpResponse): Boolean {
        return res.statusLine.statusCode == HttpStatus.SC_OK
    }

    /**
     * Wait for a certain status of a resource.
     *
     * @return either an exception or the status code of the last http response
     */
    fun waitForResource(http: TrustedHttpClient, resourceUri: URI,
                        expectedStatus: Int, timeout: Long, pollingInterval: Long): Either<Exception, Int> {
        var now = 0L
        while (true) {
            val head = HttpHead(resourceUri)
            val result = http.run<Int>(head).apply(getStatusCode)
            for (status in result.right()) {
                if (eq(status, expectedStatus) || now >= timeout) {
                    return right(status)
                } else if (now < timeout) {
                    if (!sleep(pollingInterval)) {
                        return left(Exception("Interrupted"))
                    } else {
                        now = now + pollingInterval
                    }
                }
            }
            for (e in result.left()) {
                return left(e)
            }
        }
    }
}
